### DISCUSSION TITLE
Local Storage Support Again

### DISCUSSION THREAD

**Discussion by Jaycyn** [netreconlab/Parse-Swift#183] [General]
Hello

We are evaluating platforms to migrate our 6 year old project to now that MongoDB has depreciated Realm and Atlas Sync.

During the evaluation we came across Back4App which is a strong contender for our cloud storage but - our app needs sync capability as well as local storage. The local storage would need to support ParseSwift objects, queries, create, read, update and delete.

There was a prior post / attempt at moving local storage forward but it seems to have been closed over a year ago and it appears it was aimed at the community version.

Is there something we are overlooking and if so, how does one work with local storage? We see this in the feature list

Local Storage | ⚠️ (Primitive)

And some discussion but the current/future status is unclear.

Is ParseSwift purely on online only approach?

Jay

**Comment by cbaker6** [comment]
See #69

**Comment by Jaycyn** [comment]
Thank you @cbaker6

We reviewed that conversation previously and were not fully clear on the intention or code. It was also not clear if `.current()` applied to ParseObject Structs as well.

For example, the conversation mentions the following

> All .current() objects are saved to the primitive storage such as: [ParseUser.current()](https://swiftpackageindex.com/netreconlab/parse-swift/main/documentation/parseswift/parseuser/current()), [ParseInstallation.current()](https://swiftpackageindex.com/netreconlab/parse-swift/main/documentation/parseswift/parseinstallation/current()), [ParseVersion.current()](https://swiftpackageindex.com/netreconlab/parse-swift/main/documentation/parseswift/parseversion/current()), [ParseConfig.current()](https://swiftpackageindex.com/netreconlab/parse-swift/main/documentation/parseswift/parseconfig/current()), etc.



but in practice we're not sure of the implementation: `ParseVersion.current()` works, but ParseUser, ParseConfig will not compile as they appears to be different types (ParseTypable vs ParseObject)

```
do {
    let version = try await ParseVersion.current() //works
    print("version: \(version)")
} catch {
    print("could not get version")
    assert(false)
}

do {
    let user = try await ParseUser.current() //Static member 'current' cannot be used on protocol metatype '(any ParseUser).Type'
    print("user: \(user)")
} catch {
    print("could not get user")
    assert(false)
}
```

We did review Examples of using ParseSwift with Local Storage section, but it requires an abstraction layer between whatever the local storage is (CoreData, Postgres or even json flat files) and the server.

We were hoping to utilize ParseSwift's elegance to perform queries etc against local data (like Parse used to do) and also have sync and not have to support two different databases/storage mechanisms.

Again, we are migrating from Realm and Atlas sync and were hoping ParseSwift with Back4App would fill the requirements. Bummer if not as this is a great product.

**Comment by cbaker6** [comment]
> It was also not clear if .current() applied to ParseObject Structs as well.

Almost all objects in ParseSwift are structures, I elaborate more below...

> For example, the conversation mentions the following
> 
> > All .current() objects are saved to the primitive storage such as: [ParseUser.current()](https://swiftpackageindex.com/netreconlab/parse-swift/main/documentation/parseswift/parseuser/current()), [ParseInstallation.current()](https://swiftpackageindex.com/netreconlab/parse-swift/main/documentation/parseswift/parseinstallation/current()), [ParseVersion.current()](https://swiftpackageindex.com/netreconlab/parse-swift/main/documentation/parseswift/parseversion/current()), [ParseConfig.current()](https://swiftpackageindex.com/netreconlab/parse-swift/main/documentation/parseswift/parseconfig/current()), etc.
> 
> but in practice we're not sure of the implementation: `ParseVersion.current()` works, but ParseUser, ParseConfig will not compile as they appears to be different types (ParseTypable vs ParseObject)

`ParseVersion` is a structure (same as `ParseConfig`) with a method called `.current()`. I recommend looking at the [Playgrounds](https://github.com/netreconlab/Parse-Swift/tree/main/ParseSwift.playground/Pages) as they provide many examples you can run directly from Swift Playgrounds using a local Docker instance or your Back4App server (see [Test Drive ParseSwift](https://github.com/netreconlab/Parse-Swift?tab=readme-ov-file#test-drive-parse-swift)). Also, the Playgrounds are intended to teach developers how to use ParseSwift properly.

[ParseUser](https://swiftpackageindex.com/netreconlab/parse-swift/5.11.3/documentation/parseswift/parseuser) and [ParseInstallation](https://swiftpackageindex.com/netreconlab/parse-swift/5.11.3/documentation/parseswift/parseinstallation) are [Protocols](https://developer.apple.com/videos/play/wwdc2016/419/) (see the beginning of the [README](https://github.com/netreconlab/Parse-Swift?tab=readme-ov-file) to learn about Protocol Oriented Programming (POP). You see the errors you mentioned because you never created an `object` that conforms/implements the `ParseUser` or `ParseInstallation` protocols. To implement the `User` for your type of application see the Playground example: https://github.com/netreconlab/Parse-Swift/blob/3f02f591398d6cf6467ecc8ced1e45f5714aa868/ParseSwift.playground/Pages/3%20-%20User%20-%20Sign%20Up.xcplaygroundpage/Contents.swift#L15-L45

Then you will get the `.current()` method for free (assuming you logged your `User` in): https://github.com/netreconlab/Parse-Swift/blob/3f02f591398d6cf6467ecc8ced1e45f5714aa868/ParseSwift.playground/Pages/3%20-%20User%20-%20Sign%20Up.xcplaygroundpage/Contents.swift#L72-L77

> We did review Examples of using ParseSwift with Local Storage section, but it requires an abstraction layer between whatever the local storage is (CoreData, Postgres or even json flat files) and the server.

This is true; it does require another level of abstraction/implementation, as #69 states. This is because "how" to synchronize data is left to the developer, as there are several complex solutions for various problems.

Any structure with a `.current()` is stored in the Keychain (any structure that implements the protocols mentioned earlier). Any other structure that implements the `ParseObject` protocol **IS NOT** stored in the Keychain and is instead stored in the cache. If you want to store `ParseObject''s in local storage, you will need to implement the extra abstraction and support it.

> We were hoping to utilize ParseSwift's elegance to perform queries etc against local data (like Parse used to do) and also have sync and not have to support two different databases/storage mechanisms.

ParseSwift is a very powerful SDK and is [more elegant (IMO) and functional](https://github.com/netreconlab/Parse-Swift/discussions/72) than any/all of the Parse Client SDKs (perhaps many other SDKs outside of Parse in the BaaS space, but I haven't investigated enough to make such a claim). A limitation is that it doesn't "sync" data for you, as I haven't seen a one-size-fits-all solution that has convinced me it should be implemented in the SDK. See my previous comment in #69:

> - Note that this is different from the other SDKs on purpose to provide more flexibility to the developer. You can see a discussion about the limitations of the current local storage approach in some of the other SDKs [issue 1](https://github.com/parse-community/Parse-Swift/issues/3#issuecomment-657537085) and [issue2](https://github.com/parse-community/Parse-SDK-iOS-OSX/issues/1387#issue-400547016) and why we currently believe a protocol approach would be better

**Comment by Jaycyn** [comment]
@cbaker6 Thank you so much for taking the time to provide some in-depth information.

We have an test 'ToDo' type app running with Back4App which is providing some hands on experience.

We've also looked at the Playground examples you mentioned and are considering potentially running Parse Server on the local Mac during development; but that has not gone so well.

While we have successfully installed, configured and started the Parse Server as shown here

```
jaywooten@Jay-Work ~ % parse-server --appId applicationId --masterKey masterKey --databaseURI mongodb://127.0.0.1:55419/
warn: DeprecationWarning: The Parse Server option 'encodeParseObjectInCloudFunction' default will change to 'true' in a future version.

[99268] parse-server running on http://localhost:1337/parse
```

we then ran the code in the Playground 1 - Your First Object and it crashes on the last line with no clear indication as to why (see below). We then stopped the server

` jaywooten@Jay-Work ~ % mongodb-runner stop --id=3decfacf-4065-42ec-a412-db63314a6a79`


and upon issuing the exact same command to start it again

`parse-server --appId applicationId --masterKey masterKey --databaseURI mongodb://127.0.0.1:55419/`

it error's out

```
warn: DeprecationWarning: The Parse Server option 'encodeParseObjectInCloudFunction' default will change to 'true' in a future version.
MongoServerSelectionError: connect ECONNREFUSED 127.0.0.1:55419
    at Timeout._onTimeout (/Users/jaywooten/.nvm/versions/node/v20.17.0/lib/node_modules/parse-server/node_modules/mongodb/lib/sdam/topology.js:278:38)
    at listOnTimeout (node:internal/timers:581:17)
    at process.processTimers (node:internal/timers:519:7) {
  reason: TopologyDescription {
    type: 'Unknown',
    servers: Map(1) { '127.0.0.1:55419' => [ServerDescription] },
    stale: false,
    compatible: true,
    heartbeatFrequencyMS: 10000,
    localThresholdMS: 15,
    setName: null,
    maxElectionId: null,
    maxSetVersion: null,
    commonWireVersion: 0,
    logicalSessionTimeoutMinutes: null
  },
  code: undefined,
  [Symbol(errorLabels)]: Set(0) {}
}
```

Perhaps we are doing something wrong but if the server started, nothing changed, then it should start again.

Playground crash

error: Execution was interrupted, reason: EXC_BREAKPOINT (code=1, subcode=0x19e331f28).
__lldb_expr_31/1 - Your first Object.xcplaygroundpage:316: Fatal error: Error fetching: ParseError code=104 error=objectId must not be nil

![crash](https://github.com/user-attachments/assets/c1d1f3d5-229f-43c5-bc3c-f32f3f824042)


As a last resort, we changed the `public func initializeParse` to point to our Back4App server and the Playground still crashes.

Not sure what else to do to make the Playgrounds work but we will keep working with the ToDo app we built to evaluate if ParseSwift is going to allow us to migrate from Realm.

Thanks again

Jay

**Reply by cbaker6** [reply]
The commands and errors look more like you misconfigured your parse-server. If you are trying to run locally, I suggest you follow what I mentioned in https://github.com/netreconlab/Parse-Swift?tab=readme-ov-file#test-drive-parse-swift, particularly using the docker image and commands in https://github.com/netreconlab/Parse-Swift/blob/main/CONTRIBUTING.md#swift-playgrounds as this process properly sets up a parse image on your local machine using Postgres or Mongo. If you find an issue using a known setup, be sure to report

**Reply by cbaker6** [reply]
I've tested the Playgrounds with the latest Xcode, parse-hipaa, which uses the latest parse-server, and the latest version, Parse-Swift. Everything works fine, so it's a configuration issue on your end. Some things to keep in mind when using Swift Playgrounds:
1. Playground is a synchronous environment with several asynchronous `Task` blocks. It would be best if you never attempted to run the playground past two or `Task` blocks; instead, run each `Task` block at a time and wait for it to complete. The asynchronous `Task` blocks show developers how to use Parse-Swift with async/await.
2. Never attempt to run a Playground file by going to the end and running past all of the blocks; this is mainly because of point `1` above; asynchronous code is in the Playgrounds, so the behavior is unpredictable. `Task` blocks are intended to be run one after another, not run in parallel
3. The Playgrounds are the best place to test out Parse-Swift as they contain the most examples of how to use the latest version of the SDK

You are combining points 1 and 2 above based on the errors you posted.

You can easily have Pare-Swift connect any working Parse-Server like Back4App by changing the configuration as long as the server is configured correctly. To point Parse-Swift Playgrounds to your server, change (also mentioned in [CONTRIBUTING.md](https://github.com/netreconlab/Parse-Swift/blob/main/CONTRIBUTING.md#swift-playgrounds)) https://github.com/netreconlab/Parse-Swift/blob/3f02f591398d6cf6467ecc8ced1e45f5714aa868/ParseSwift.playground/Sources/Common.swift#L4-L12

**Comment by Jaycyn** [comment]
Thank you.

As mentioned, we have the Parse Server successfully running (we think) and we also pointed the initialization code to our Back4App server with the same result.

We assume since we can access the GUI, the server is running. Is that correct?

![Server](https://github.com/user-attachments/assets/31695d6d-004d-497f-88bf-25ae5fead279)

We opened XCode and in navigated to the 1 - Getting Started and ran the Playground - here's the result. 144 errors due to Optional Data -> String Conversion Violation:. However, I believe this is a fault in SwiftLint default configuration but don't want to change anything without knowing the correct path.

![Code](https://github.com/user-attachments/assets/347be2ba-25d5-4d2f-acc6-d9c7cd057786)

We then thought there maybe an update available via SPM so upon checking that we receive:

![No Respository](https://github.com/user-attachments/assets/759b7642-1fe1-46af-9006-8cca7bba63cd)

**Reply by cbaker6** [reply]
> We opened XCode and in navigated to the 1 - Getting Started and ran the Playground - here's the result. 144 errors due to Optional Data -> String Conversion Violation:. However, I believe this is a fault in SwiftLint default configuration but don't want to change anything without knowing the correct path.

The Swiftlint error is different from what you mentioned before. [SwiftLint](https://github.com/realm/SwiftLint/releases/tag/0.57.0), recently made an update, so you must be using the latest version if you are seeing this error. To fix, add the following line to the respective file, and all of the compile errors will go away https://github.com/netreconlab/Parse-Swift/blob/093542eb2b96f034ea3efe758790580a7e09aada/.swiftlint.yml#L6

> We then thought there maybe an update available via SPM so upon checking that we receive:

As for the SPM error, you should not need to do anything with SPM if you forked https://github.com/netreconlab/Parse-Swift

> We assume since we can access the GUI, the server is running. Is that correct?

Maybe, you may have to click more through the dashboard to ensure all the default Classes are showing. In addition, you would need to go through the logs to verify. In addition, this doesn't mean it's necessarily configured correctly to handle the Parse-Swift SDK and/or the Parse Swift Playgrounds. The docker image I provided I know for a fact is configured correctly, there's no way for me to tell through screenshots that you configured your manual server correctly. Note that developers misconfigure parse servers all of the time and think it's the SDK is the issue.

Lastly, from your Parse Dashboard screenshot, you are running an old version of the Parse Server. The latest Parse Server is version [7.3.0](https://github.com/parse-community/parse-server/releases/tag/7.3.0). The version in your screenshot is no longer supported by Parse and is not in [LTS](https://github.com/parse-community/parse-server?tab=readme-ov-file#long-term-support). Though Parse-Swift works with old versions of the server, it stays in line with the latest version and features, so it's possible you could be testing against a server with bugs, security flaws, and/or missing features. Many times the missing features may be mentioned in the documentation.

**Comment by Jaycyn** [comment]
Thanks for the response and patience.

For clarity, we don't have a manual server - we used the docker link (the same one you provided) and went through that process. Not sure how we would end up with an old version of the Parse Server though. We were trying to manually start it which was referenced above. I am not seeing a --version command to determine what version installed. Perhaps there's another way?

We had also pointed the func initialize to our current Back4App server which is working and received the same error. Here's what Docker looks like - we can start and stop parse-hipaa-parse-swift container.

![Docker](https://github.com/user-attachments/assets/27b7b7fb-7093-4694-b469-40aef9c3ea3e)

We had quite a few issues with SwiftLint and Realm previously so it was not installed. We installed it, added the line you mentioned

![SwiftLint](https://github.com/user-attachments/assets/28c6d096-6cf5-449f-a215-28deadcd534a)

And then back to the same error we previously mentioned:

![Same error](https://github.com/user-attachments/assets/698ae9fb-dcaf-4cdf-8fe4-39f5a2e3e987)

To summarize

Docker is installed and running the Parse-Server. 
In the Parse Dashboard web app, it shows v 5.4.0
Default classes are showing in Dashboard
Dashboard logs only show

- 2024-10-03T20:45:43.533Z - Invalid Option Keys Found: enableSchemaHooks, liveQuery.websocketTimeout, liveQuery.cacheTimeout, graphMountPath
- 2024-10-03T20:45:43.531Z - DeprecationWarning: The Parse Server option 'encodeParseObjectInCloudFunction' default will change to 'true' in a future version.

We can navigate around and see _Role and _User (we assume those are the Default Classes you referenced)
No other changes were made

![Parse Dashboard](https://github.com/user-attachments/assets/923a8491-3bca-4e57-b05b-e951221204ab)

**Reply by cbaker6** [reply]
I'm guessing you are still doing what I mentioned you shouldn't do in https://github.com/netreconlab/Parse-Swift/discussions/183#discussioncomment-10826250; in particular points 1 and 2.

The Playgrounds is a synchronous environment with a bunch of asynchronous code; you **should not** jump to the end of the file and press play. What happens if you go to `line 35` and hit play; Don't do anything else after, but wait for the console at the bottom to output information: 
<img width="858" alt="image" src="https://github.com/user-attachments/assets/963017c2-e0fe-4b67-be1e-52ff65181594">

You should see the following in the Playground console output:

```bash
Current Swift SDK version is "5.11.3"
Server health is: ok
```

Once the console output appears and shows the version, etc. Then go to `line 166` and hit play (**do not restart the Playground** or else you will need to start from the beginning); stop and wait for output again: 
<img width="864" alt="image" src="https://github.com/user-attachments/assets/a9b799eb-7f00-44e1-9e71-249f3d16d3cc">

Now you should see the following in the console output:

```bash
Saved "GameScore" with the following info:
    GameScore ({"createdAt":{"__type":"Date","iso":"2024-10-03T22:57:43.764Z"},"objectId":"d1rHxZzyDNyjsGyEbaagJFAJXiQ38akj","points":10,"updatedAt":{"__type":"Date","iso":"2024-10-03T22:57:43.764Z"}})
    Updated "GameScore" with the following info:
    GameScore ({"createdAt":{"__type":"Date","iso":"2024-10-03T22:57:43.764Z"},"objectId":"d1rHxZzyDNyjsGyEbaagJFAJXiQ38akj","points":200,"updatedAt":{"__type":"Date","iso":"2024-10-03T22:57:43.791Z"}})
```

If the following works, then you should be stepping through the rest of the Playgrounds in a similar fashion.

**Comment by Jaycyn** [comment]
As an update: We are very familiar with XCode, Playgrounds etc and have been using those tools since they were released and have several apps in pre-production so we know to no jump to the end to run the code (not to date myself, but i've been coding for 40+ years so I kinda know my way around).

One ongoing issue is that Docker Desktop is giving us a lot of trouble; once it's used and then quit, all of the background activities continue to run and must be force quit with Activity Monitor for it to be usable again. That leads to, what we guess, is the parse server not fully starting. We removed all versions of parse, parse-server as well as the quick-start hippa server and then used GitHub desktop to refresh everything.

And it's almost working.

Following your instructions above, up to line 35 works and outputs to console. But then, without restarting playground, it crashes before line 166

Current Swift SDK version is "5.11.3"
Server health is: ok
__lldb_expr_25/1 - Your first Object.xcplaygroundpage:163: Fatal error: Error saving: ParseError code=-1 error=unauthorized

![Line166_crash](https://github.com/user-attachments/assets/0b55a6bc-92cd-4c5c-af36-37befc5ac560)

**Reply by cbaker6** [reply]
> One ongoing issue is that Docker Desktop is giving us a lot of trouble; once it's used and then quit, all of the background activities continue to run and must be force quit with Activity Monitor for it to be usable again. That leads to, what we guess, is the parse server not fully starting. We removed all versions of parse, parse-server as well as the quick-start hippa server and then used GitHub desktop to refresh everything.

The docker setup is intended to be straightforward. Note, there have many developers whether in industry or undergraduates in my classes that use the docker images/[documentation](https://github.com/netreconlab/parse-hipaa/tree/parse-swift?tab=readme-ov-file#local-using-docker-image-with-postgres-or-mongo), so I don't quite understand the issues you are running into. Nonetheless, I recommend the following to attempt to remedy your issues:

1. Stop all/any Parse related containers in your Docker Desktop
2. Delete all of the Parse related containers from Docker Desktop (Be sure to allow any permissions Docker has asked for when you installed)
3. Use `terminal` and change directories to the `parse-hipaa` folder
4. In terminal, type `git checkout parse-swift` (if you made any changes to any files in this branch in the past, delete/revert them (or reclone the branch). You should only be using the raw code I provided in this branch. You can make changes after you ensure the base setup works properly)
5. In terminal, type `docker images`, this should should show a list of images
6. In terminal, delete every Parse related and Postgres/Mongo related images using its respective `IMAGE ID` using, `docker image rm`. For example if two ID's exist: `66e11c00515e`, `8ccf12182d23`, use `docker image rm 66e11c00515e 8ccf12182d23`
7. In terminal, type `docker-compose up -d`, if you receive an error about `docker-compose` not being available, it's probably because your computer has the latest version of docker, you can try `docker compose up -d` instead). You should see the following (all containers should be green): 
<img width="1240" alt="image" src="https://github.com/user-attachments/assets/8e34a011-3f6b-4393-ae02-6ffd66ff566f">
8. Follow the [directions](https://github.com/netreconlab/parse-hipaa/tree/parse-swift?tab=readme-ov-file#local-docker-or-singularity) here to ensure the Dashboard is connected to your server
9. Goto your Parse-Swift cloned Xcode Project, be sure to revert any changes you made to your branch, it should be fresh clone with no changes to the `initializeParse()` method as it's configured by default to connect to a proper working Docker image 
10. Now repeat the steps I mentioned in https://github.com/netreconlab/Parse-Swift/discussions/183#discussioncomment-10838148. Everything should work (I just completed the whole process from scratch myself, I ran into zero issues)
11. Once you are finished testing Parse-Swift with Docker, type `docker compose down` in the terminal

**Comment by Jaycyn** [comment]
Those last steps were incredibly helpful. I believe we now have it working (finally), and the Playgrounds are behaving normally.

I think the above mentioned issue with Docker Desktop processes may be part of the issue as everything as working correctly from the command line. We are working on setting this up on two employee workstations and we will see how it goes. Will report back.

Again, thank  you for your patience and guidance.
